### DISCUSSION TITLE
Extend coupons, affiliate links from third party auto-apply coupon for discount

### DISCUSSION THREAD

**Discussion by neekfenwick** [zencart/zencart#5856] [Ideas]
I've had a request to allow an affiliate link to bring a potential customer to our site with a discount applied, like "Use this link to go to ourstore.com and get 10% off your purchase".  I have a notion that the 'coupon' feature of Zen Cart could be used for this by introducing a new way of applying a coupon that doesn't require the user to type in the coupon code.

Note: This is different to the Snap Affiliates addon https://www.zen-cart.com/downloads.php?do=file&id=1635 which would earn commissions for the affiliate website.  This feature is about giving the customer visiting our shop the discount.  It is similar in that the referer header would be checked to confirm where the visitor is coming from.

A coupon could be configured so that visitors from e.g. Bob's Bits website get a discount, the init code could look up the relevant coupon, check it hasn't been used by the current customer before (I expect using the usual coupon.uses_per_user=1), and apply the coupon automatically.  Once applied it would be like any other coupon.

Checking the visitor could either be by the coupon knowing the Referer header to expect e.g. bobsbits.com, and/or a query param the URL must contain e.g. "?affiliate_code=bobsbits".

I'm not very familiar with the coupon system, how it works for guests and the transition to being logged on for checkout, etc.  Wanted to float the idea and see if people think it's good, or would be better handled another way or by an existing addon I'm unaware of.

**Comment by scottcwilson** [comment]
Go for it.  I like this idea.

**Comment by neekfenwick** [comment]
@scottcwilson I've had a go at this on my branch https://github.com/neekfenwick/zencart/tree/coupon-referer-discussion-5856 .. I'm not sure how to produce an easy to read diff on github without creating a PR.  Summary:

1. upgrade SQL addition to create column `referrer` on table `coupons`.
2. admin page changes for new 'Referrer' field (view/edit/save/etc).
3. catalog init script to sniff HTTP_REFERER and enable matching coupon, if any.

The biggest change is in the new init script https://github.com/neekfenwick/zencart/blob/coupon-referer-discussion-5856/includes/init_includes/init_coupons.php and there are a few things I'm not comfortable with, I'm hoping someone can help me out:

1. The `ot_coupon` class is normally instantiated in the checkout process from inside `order_totals`, which has a list of all order totals modules, loads their language files, and instantiates them all.  So, I've had to load the language strings and instantiate `ot_coupon` in order to check if it is even enabled.  This is now done in `init_coupons` for every single page load, which seems heavy handed to me.  I could simplify by directly sniffing the `coupons` table and not instantiate `ot_coupon` at all, but this would break encapsulation i.e. have coupon logic done outside of `ot_coupon`.
2. `init_coupon` is at `config.core` breakpoint 138, quite late in the load sequence, because it has to come after `init_templates` for `$languageLoader` and after `$messageStack`.  I guess that's fine.
3. The 'success' messageStack uses pre-existing `TEXT_VALID_COUPON` which is not very informative.  In order to make a more descriptive message, e.g. naming the coupon that has been auto-added, I'd need a new lang string.  I've added a couple in admin for the new on-page fields, but so far no changes to catalog lang strings.

Comments would be most welcome.

**Comment by drbyte** [comment]
This is how I did it a long time back in a much older ZC version: https://gist.github.com/drbyte/d503661e2b91353e82f8790a2e982cfb
The objective wasn't so much to auto-apply an actual discount, but merely to use the coupon-tracking feature to track who referred the customer. But if the coupon code you use for tracking is set to a non-zero discount then of course it will get applied.

Here are some forum threads where similar things were discussed, with varying approaches taken:
https://www.zen-cart.com/showthread.php?173762-Redeeming-coupon-code-from-URL
https://www.zen-cart.com/showthread.php?42386-Coupon-Code-in-URL

Here's an old plugin that claims to do it ... but they added a lotta extra code ... https://www.zen-cart.com/downloads.php?do=file&id=1727

**Reply by neekfenwick** [reply]
Right, your gist applies `$_SESSION['cc_id']` whether the `ot_coupon` module is enabled or not, which seems low impact since it would be ignored, but you don't display any `$messageStack` indicator to the customer that anything has happened.  The two main problems I ran into as I described above are 1/ using `ot_coupon` to validate and report on the coupon attempt 2/ getting language strings loaded for the `ot_coupon` module to function (e.g. its ctor uses a lang string immediately and crashes if lang is not loaded).

On your tip, I read those two forum threads, all problems faced are certainly similar and face the same problems as I describe above.  I feel my solution using an init script and testing the requested coupon against `ot_coupon`'s validation is the right way to go, I just don't like the extra code and init work I had to do to get there :)  It looks like the cost of doing business with `ot_coupon` is to do all that work up front.

The addon you linked to is certainly a lot of code.  It looks like it does a bunch of validation that I think is core ZC now, for example they sniff the cart for valid products which is now handled by calling `$ot_coupon->performValidations($coupon_code)` in my `init_coupons.php`.  So, I think my attempt subsumes that addon and would make it core.

**Comment by neekfenwick** [comment]
I'm intending to come back to this, when I find time.

To minimise work done in the init_coupons.php file (done for every page load), it seems better to directly query the coupons database table to see if any match the current referer header before initialising any lang strings or `ot_coupon` itself.  This would move some coupon logic outside of `ot_coupon`.  If a referer is found, then init_coupons.php can continue with what's in my current work in progress branch .. load the lang strings and instantiate ot_coupon, then invoke its validation code to check the coupon is actually valid (in date, etc).

I'm not clear from the previous discussion whether this would be rejected on some other grounds, please feedback here if I should not bother with any of this.

**Reply by drbyte** [reply]
Yes, that's a sensible approach. No need to fire up ot_coupon unless there's a described need for it. Just query for the referrer if it's even present.

I'd support having init_coupon also fire up ot_coupon if an agreed-upon GET param is present. ie: iterating on the concept I'd posted, but addressing the validation lookups so that messages can be displayed about it being applied or only partially-valid for cart contents.

**Comment by drbyte** [comment]
Thinking through the various factors here, the one that sticks out after seeing how much code is needed just to add a new field and allow data-entry and data-lookups on it, is the awareness that a lot of CMS sites use [`rel="no_referrer"` attributes](https://www.elegantthemes.com/blog/wordpress/rel-noopener-noreferrer-nofollow) on links posted by "visitors" (but can probably be bypassed for authorized users or admins). While it helps focus SEO a bunch, it breaks link-click-based referral/affiliate tracking. 
This is one reason why the code I posted about simply "auto-applied the coupon if found in the `GET` params of the URL" as its sole use-case.

**Reply by neekfenwick** [reply]
You're right, but this feature isn't aimed at links posted by visitors, it's for when a blog's main content contains (usually paid for) links.  Say I pay $300 to a blogger to host a blog post about my stuff that links to my shop, she would state "click here for a 10% discount" and I want people clicking that link to get the coupon auto-applied.  The link wouldn't have `rel=no_referrer`.

Your second point is valid though, this would be a good point to check `$_GET` and auto-apply a coupon based on it.  I have pushed https://github.com/neekfenwick/zencart/blob/653b382854838d6bda36641c776a4325edf2dd16/includes/init_includes/init_coupons.php#L45 with two checks, one for referrer, and one for `$_GET['coupon_code']`, applying whichever is found.

What's the right way to sanitise $_GET these days?  I've used `zen_db_prepare_input($_GET['coupon_code'])`.

**Comment by neekfenwick** [comment]
Please see PR https://github.com/zencart/zencart/pull/6178
Apparently a build is failing (https://github.com/zencart/zencart/actions/runs/7708525243/job/21007789361?pr=6178) but I've no idea why, it looks like the verbose error output is disabled during github builds.  Something to do with the new `coupons` db field?  I don't know how to run the same build step locally, running `php vendor/bin/phpunit  --verbose --testsuite Feature --printer 'Sempro\PHPUnitPrettyPrinter\PrettyPrinterForPhpUnit9'` gives:

    PHP Fatal error:  Uncaught Error: Class "Symfony\Bundle\FrameworkBundle\Test\WebTestCase" not found in /home/neek/workspace/zen-cart/zencart/not_for_release/testFramework/Support/zcFeatureTestCase.php:20

**Reply by drbyte** [reply]
The SQL for fresh installs was missing the new field.

**Comment by drbyte** [comment]
Merged.
